### PR TITLE
fixed commit

### PR DESCRIPTION
**Pull request by rjtch**
installed and caches steps for rust and rust

### CODE REVIEW

**Comment by rjtch** [comment]
This is in WiP and will be able to be merged soon.

File: .github/workflows/linter.yaml [hunk 1]
```diff
@@ -5,46 +5,71 @@ permissions:
 
 on:
   push:
+    paths:
+      - ".github/workflows/linter.yml"
+      - "**.rs"
+      - "**.toml"
+      - "**/Cargo.lock"
+      - "**/Makefile"
```
Comment (afrittoli) [inline .github/workflows/linter.yaml:13]: Any reason for adding `paths` here? 
One of the linters you added checks the format of commit messages, and that should run on every commit.

File: .github/workflows/linter.yaml [hunk 2]
```diff
@@ -5,46 +5,71 @@ permissions:
...
     branches: [ "main" ]
   pull_request:
+    paths:
```
Comment (afrittoli) [inline .github/workflows/linter.yaml:16]: Ditto

File: .github/workflows/linter.yaml [hunk 3]
```diff
@@ -5,46 +5,71 @@ permissions:
...
-concurrency:
-  group: ${{ github.workflow }}-${{ github.ref }}-${{ github.event_name }}
-  cancel-in-progress: true
-
-defaults:
-  run:
-    shell: bash
+# concurrency:
+#   group: ${{ github.workflow }}-${{ github.ref }}-${{ github.event_name }}
+#   cancel-in-progress: true
```
Comment (afrittoli) [inline .github/workflows/linter.yaml:30]: NIT: Please do not leave commented-out sections - we can add this back in future if needed

File: .github/workflows/linter.yaml [hunk 4]
```diff
@@ -5,46 +5,71 @@ permissions:
...
 
 jobs:
   lint_commits:
     name: Lint Commit Messages
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-22.04
```
Comment (afrittoli) [inline .github/workflows/linter.yaml:35]: +1

File: .github/workflows/linter.yaml [hunk 5]
```diff
@@ -5,46 +5,71 @@ permissions:
...
     steps:
       - name: Checkout repository
         uses: actions/checkout@v4
       - name: Check Commit Lint
-        uses: wagoid/commitlint-github-action@v5.0.0
+        uses: wagoid/commitlint-github-action@v5.4.5
 
   lint_check:
-    name: Rust - lint_${{ matrix.lint_projects }}
+    name: Rust - lint_all
```
Comment (afrittoli) [inline .github/workflows/linter.yaml:43]: Why this change? I think the original name was better.

File: .github/workflows/linter.yaml [hunk 6]
```diff
@@ -5,46 +5,71 @@ permissions:
...
       fail-fast: false
       matrix:
         lint_projects:
           - cargo_fmt_check
           - cargo_clippy
-          - cargo_deny
+#          - cargo_deny
           - cargo_toml_files
     steps:
-      - name: Checkout repository
```
Comment (afrittoli) [inline .github/workflows/linter.yaml:46]: NIT: I do prefer having a meaningful name for steps, it makes the workflow more readable

File: .github/workflows/linter.yaml [hunk 7]
```diff
@@ -5,46 +5,71 @@ permissions:
...
+        restore-keys: |
+             ${{ runner.os }}-sdk-rust-
+      - name: Check cargo version
+        run: cargo --version
+      - name: Format
+        run: cargo fmt --all -- --check
+      - name: Clippy
+        run: cargo clippy --verbose -- -D warnings
+      - name: Check Cargo.toml files
+        run: cargo check --workspace --all-targets --all-features
```
Comment (afrittoli) [inline .github/workflows/linter.yaml:72]: You added various checks as steps in one job, but you still have the matrix, so now these checks will be executed three times. 
If you prefer one job (to avoid installing things multiple times) it's fine, but please remove the matrix then.

File: .github/workflows/linter.yaml [hunk 8]
```diff
@@ -5,46 +5,71 @@ permissions:
...
+      - uses: taiki-e/install-action@protoc
+      - uses: dtolnay/rust-toolchain@beta
+        with:
+          profile: minimal
+          toolchain: nightly
+          components: rustfmt, clippy
+      - uses: Swatinem/rust-cache@v2
+        key: ${{ runner.os }}-sdk-rust-${{ hashFiles('**/Cargo.lock') }}
+        restore-keys: |
+             ${{ runner.os }}-sdk-rust-
```
Comment (afrittoli) [inline .github/workflows/linter.yaml:64]: Could you add a name to each of this steps that explains their function please?

File: .github/workflows/linter.yaml (hunk 7 above)
Comment (afrittoli) [inline .github/workflows/linter.yaml:72]: Also, why are this not using the make file anymore?

File: tools/cargo-deny/deny.toml [hunk 9]
```diff
@@ -10,17 +10,17 @@ unlicensed = "deny"
 copyleft = "deny"
 confidence-threshold = 0.95
 allow = [
-    "MIT",
     "Apache-2.0",
+    "MIT",
+]
+exceptions = [
+    { name = "cloudabi", allow = ["BSD-2-Clause"] },
```
Comment (afrittoli) [inline tools/cargo-deny/deny.toml:17]: Is this the license in question? https://opensource.org/license/bsdpluspatent/ 
If so, it's an OSI approved license, so we don't need an exception, we can add it to the allow list.

File: Cargo.toml [hunk 10]
```diff
@@ -1,4 +1,5 @@
 [workspace]
+resolver = "2"
```
Comment (afrittoli) [inline Cargo.toml:2]: Including the 2021 edition would imply this: https://doc.rust-lang.org/edition-guide/rust-2021/default-cargo-resolver.html

**Comment by rjtch** [comment]
Sorrry @afrittoli i will rework it i changed everything because of this message ´taiki-e/install-action@protoc, dtolnay/rust-toolchain@beta, and swatinem/rust-cache@v2 are not allowed to be used in cdevents/sdk-rust. Actions in this workflow must be: within a repository owned by cdevents, created by GitHub, or matching the following: wagoid/commitlint-github-action@*.´ https://github.com/cdevents/sdk-rust/actions/runs/7576295830.

**Comment by afrittoli** [comment]
> Sorrry @afrittoli i will rework it i changed everything because of this message ´taiki-e/install-action@protoc, dtolnay/rust-toolchain@beta, and swatinem/rust-cache@v2 are not allowed to be used in cdevents/sdk-rust. Actions in this workflow must be: within a repository owned by cdevents, created by GitHub, or matching the following: wagoid/commitlint-github-action@*.´ https://github.com/cdevents/sdk-rust/actions/runs/7576295830.

I can enable any action we need in the repo config, by default non-GitHub-maintained actions are disallowed for security reasons.

File: .github/workflows/linter.yaml [hunk 11]
```diff
@@ -17,34 +19,54 @@ concurrency:
...
+      - name: Run the checkout command
         uses: actions/checkout@v4
+      - name: Install rust toolchain
+        uses: actions-rust-lang/setup-rust-toolchain@v1
+        with:
+          profile: minimal
+          toolchain: nightly
+          components: clippy, rustfmt
+      - name: Install development tools
+        uses: taiki-e/install-action@protoc
```
Comment (afrittoli) [inline .github/workflows/linter.yaml:54]: I think this should be `taiki-e/install-action@v2`
Do we need `protoc`? If we do, you can add it to the list of tools

Comment (rjtch) [reply .github/workflows/linter.yaml:54]: yes but taiki-e/install-action@v2 is not accepted! could you please remove protoc and add ..@v2 instead?

**Comment by afrittoli** [comment]
Thanks for the latest updates, it's looking good!

I think for the commit check to pass it would be easier if you squashed all commits to one, and updated the commit message. It would be nice if you could set the PT title and description to something with a bit more context too 🙏

**Comment by rjtch** [comment]
> Thanks for the latest updates, it's looking good!
> 
> I think for the commit check to pass it would be easier if you squashed all commits to one, and updated the commit message. It would be nice if you could set the PT title and description to something with a bit more context too 🙏

yes sure i will

File: tools/cargo-deny/deny.toml [hunk 12]
```diff
@@ -10,17 +10,16 @@ unlicensed = "deny"
 copyleft = "deny"
 confidence-threshold = 0.95
 allow = [
-    "MIT",
-    "Apache-2.0",
+  "Apache-2.0",
+  "MIT",
+  "BSD-2-Clause",
```
Comment (davidB) [inline tools/cargo-deny/deny.toml:15]: For my PR, more licenses will be needed (like `CC0-1.0`) some by lib used by the generator, some by delivered sdk.

Do you think, we should use the same rules for both (as the generator is not distributed)?

Comment (rjtch) [reply tools/cargo-deny/deny.toml]: yes but i don't know actually how the cd-event team would like to manage linsences but for me it is not a problem.

File: tools/dprint/dprint.json [hunk 13]
```diff
@@ -1,6 +1,8 @@
 {
   "includes": [
-    "**/*.toml"
+    "**/*.toml",
+    "**/*.yaml",
+    "**/*.rs"
```
[the remark this answers is not in the document]
Comment (rjtch) [reply tools/dprint/dprint.json:5]: my problem with mega-superlinter is the size of the image which are too big. One thing we could do would be to use prettier as pre-commit for linter all files and then remove this.

File: .github/workflows/linter.yaml [hunk 14]
```diff
@@ -35,38 +37,36 @@ jobs:
...
         with:
           profile: minimal
           toolchain: nightly
           components: clippy, rustfmt
       - name: Install development tools
-        uses: taiki-e/install-action@protoc
+        uses: taiki-e/install-action@v2
         with:
-          tool: just, cargo-deny, dprint
+          tool: cargo-deny, dprint
```
[the remark this answers is not in the document]
Comment (rjtch) [reply .github/workflows/linter.yaml]: By default cd-event resctrict the use of public actions so they have to be added after some inspection. I use this to have as small image size as possible.
